Src/CSharpier.Cli/FormattingCache.cs

```diff
@@ -0,0 +1,144 @@
+        CancellationToken cancellationToken
+    )
+    {
+        if (!commandLineOptions.Cache)
+        {
+            return NullCache;
+        }
+
+        // TODO document how to delete it
+        var cacheFile = Path.Combine(
+            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
+            "CSharpier",
+            ".formattingCache"
+        );
+
+        ConcurrentDictionary<string, string> cacheDictionary;
+        if (!File.Exists(cacheFile))
+        {
+            cacheDictionary = new ConcurrentDictionary<string, string>();
+        }
+        else
+        {
+            cacheDictionary =
+                JsonSerializer.Deserialize<ConcurrentDictionary<string, string>>(
+                    await File.ReadAllTextAsync(cacheFile, cancellationToken)
+                ) ?? new();
+        }
+
+        return new FormattingCache(printerOptions, cacheFile, cacheDictionary, fileSystem);
+    }
+
+    private class FormattingCache : IFormattingCache
+    {
+        private readonly string optionsHash;
+        private readonly string cacheFile;
+        private readonly ConcurrentDictionary<string, string> cacheDictionary;
+        private readonly IFileSystem fileSystem;
+
+        public FormattingCache(
+            PrinterOptions printerOptions,
+            string cacheFile,
+            ConcurrentDictionary<string, string> cacheDictionary,
+            IFileSystem fileSystem
+        )
+        {
+            this.optionsHash = GetOptionsHash(printerOptions);
+            this.cacheFile = cacheFile;
+            this.cacheDictionary = cacheDictionary;
+            this.fileSystem = fileSystem;
+        }
+
+        public bool CanSkipFormatting(FileToFormatInfo fileToFormatInfo)
+        {
+            var currentHash = Hash(fileToFormatInfo.FileContents) + this.optionsHash;
+            if (this.cacheDictionary.TryGetValue(fileToFormatInfo.Path, out var cachedHash))
+            {
+                DebugLogger.Log(fileToFormatInfo.Path + " " + currentHash + " " + cachedHash);
+                if (currentHash == cachedHash)
+                {
+                    return true;
+                }
+
+                this.cacheDictionary.TryRemove(fileToFormatInfo.Path, out _);
+            }
+
+            return false;
+        }
+
+        public void CacheResult(string code, FileToFormatInfo fileToFormatInfo)
+        {
+            this.cacheDictionary[fileToFormatInfo.Path] = Hash(code) + this.optionsHash;
+        }
+
+        private static string GetOptionsHash(PrinterOptions printerOptions)
+        {
+            var csharpierVersion = typeof(FormattingCache).Assembly.GetName().Version;
+            return Hash($"{csharpierVersion}_${JsonSerializer.Serialize(printerOptions)}");
+        }
+
+        private static string Hash(string input)
+        {
+            using var md5 = MD5.Create();
+            var result = md5.ComputeHash(Encoding.ASCII.GetBytes(input));
```

Comment:
MD5 is kind of outdated and if you enable code analysis they will warn you when they see it.
Better of using SHA-1 or one of its variants from the get go.

Reply:
I assumed MD5 would be faster even if it is outdated, but it turns out SHA1 is actually a bit faster. Or using the new XxHash32 is even faster with a smaller cache size!

Working on a repo with 11k files, non-scientific benchmark (ran it a few times and did the average in my head), with everything already formatted
md5 3.9s 2mb file
sha1 3.8s 2.2mb file
xxh32 3.4s 1.5mb file

It does require using a preview version of System.IO.Hashing. There are no open issues related to xxh32 and the recent history is pretty quiet - https://github.com/dotnet/runtime/commits/main/src/libraries/System.IO.Hashing